src/js/ProviderManager.js

```diff
@@ -0,0 +1,45 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+/*globals async, DirectoryLinksProvider, Links, PlacesProvider, gUserDatabase, gPinnedLinks, gBlockedLinks*/
+
+/**
+ * Singleton that handles link providers.
+ */
+
+ "use strict";
+
+const ProviderManager = {
+  init: async(function*(gMockFetch) {
```

Comment:
I don't think it's necessary to monkey patch anything (although, it's fun, because :see_no_evil: and :hear_no_evil: and :speak_no_evil: are fun!). All we want to do is throw bogus responses at this thing. That should be trivial to do. 

I'll come up with something after I finish this round.